### ISSUE TITLE
Convert existing native histogram unit tests to the PromQL testing framework

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#13563]
### Proposal

The PromQL testing framework has been upgraded to support native histograms, but we still have a lot of histogram tests written as code level unit tests that we can now convert into much simpler tests within the PromQL testing framework.

In other words: More tests need to be added to https://github.com/prometheus/prometheus/blob/main/promql/testdata/histograms.test that do exactly the same as the following unit tests (which we can then delete):

- [x] [TestNativeHistogramRate](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3155)
- [x] [TestNativeFloatHistogramRate](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3243C6-L3243C34)
- [x] [TestNativeHistogram_HistogramCountAndSum](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3284C6-L3284C46)
- [x] [TestNativeHistogram_HistogramStdDevVar](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3363C6-L3363C44)
- [x] [TestNativeHistogram_HistogramQuantile](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3522C6-L3522C43)
- [x] [TestNativeHistogram_HistogramFraction](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3758C6-L3758C43)
- [ ] [TestNativeHistogram_Sum_Count_Add_AvgOperator](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L4193C6-L4193C51)
- [ ] [TestNativeHistogram_SubOperator](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L4388C6-L4388C37)
- [ ] [TestNativeHistogram_MulDivOperator](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L4623C6-L4623C40)

**Comment by beorn7** [comment]
@zedGGs might be interested in this.

I recommend to pick a single unit test first and convert it to get familiar with the procedure.

**Comment by NeerajGartia21** [comment]
I would like to work on this one. Can you please assign it to me?

**Comment by beorn7** [comment]
Thanks. I have assigned this to you. Let me know if you have any questions.

**Comment by NeerajGartia21** [comment]
Thanks for assigning it to me! 
I'll first try to convert [TestNativeHistogramRate](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L3155) and after verification I'll move with other tests.

**Comment by NeerajGartia21** [comment]
@beorn7 more precisely the test are to be added to https://github.com/prometheus/prometheus/blob/main/promql/testdata/native_histograms.test right?

**Comment by beorn7** [comment]
Yes. And then delete the corresponding Go unit test code.

**Comment by beorn7** [comment]
Three tests are left to convert, see TODO list at the top. @NeerajGartia21 are you planning to convert them, too?

**Comment by NeerajGartia21** [comment]
Yes, @beorn7. I was busy because of my onboarding process but I'll convert it ASAP.

**Comment by beorn7** [comment]
@NeerajGartia21 are you still planning to work on this?

**Comment by fatsheep9146** [comment]
@beorn7 I would like to take this issue, if @NeerajGartia21 do not have time now.

**Comment by beorn7** [comment]
With no message from @NeerajGartia21 , assigning this to @fatsheep9146 .

**Comment by fatsheep9146** [comment]
When I tried to convert `TestNativeHistogram_Sum_Count_Add_AvgOperator` into framework, I came up with two problems

1. `sum` operator 

when evaluating the expression `sum(histogram_sum)`, the result we get just looks the same as expected, but the test are still failed with following error.

> Error: Received unexpected error:
> error in eval sum(histogram_sum) (line 9): expected {{count:107 sum:4691.2 z_bucket:14 z_bucket_w:0.001 buckets:[3 8 2 5 3 2 2] n_buckets:[2 6 8 4 15 9 0 0 0 10 10 4]}} for {} but got {{count:107 sum:4691.2 z_bucket:14 z_bucket_w:0.001 buckets:[3 8 2 5 3 2 2] n_buckets:[2 6 8 4 15 9 0 0 0 10 10 4]}}

After digging into this, I found the different between actual result and expected result happened on negative spans, the expected negative spans will be Compact [here](https://github.com/prometheus/prometheus/blob/e92a18b6ce295b99e1c089d6e2bc62fbe3f73869/promql/promqltest/test.go#L807) before comparing with actual result. So the expected negative spans is `[{0 6} {3 3}]`, and the actual negative spans is `[{0 12}]`. 

So I think the actual results should also be compacted with `Compact(0)` before comparing.


2. `avg`, `sum_over_time` and `avg_over_time` operator

Those three operator sometimes will produces unexpected float number for native histogram's sum or buckets, for example, 
> error in eval avg(histogram_sum) (line 20): expected {{count:26.75 sum:1172.8 z_bucket:3.5 z_bucket_w:0.001 buckets:[0.75 2 0.5 1.25 0.75 0.5 0.5] n_buckets:[0.5 1.5 2 1 3.75 2.25 0 0 0 2.5 2.5 1]}} for {} but got {{count:26.75 sum:1172.7999999999997 z_bucket:3.5 z_bucket_w:0.001 buckets:[0.75 2 0.5 1.25 0.75 0.5 0.5] n_buckets:[0.5 1.5 2 1 3.7499999999999996 2.25 0 0 0 2.5 2.5 1]}}

The expected sum should be 1172.8, not 1172.7999999997, this situation does not necessarily occur; it only happens occasionally. I still do not know why. But I thinks we could enhance the by comparing the field of float type with [Almost.Equal()](https://github.com/prometheus/prometheus/blob/e92a18b6ce295b99e1c089d6e2bc62fbe3f73869/util/almost/almost.go#L22), which could toleration some tiny diff between two floats.

I already make some test for these two ways of enhancing the comparing logic of two native histogram, it worked.

@beorn7

**Comment by beorn7** [comment]
> So I think the actual results should also be compacted with Compact(0) before comparing.

Sounds like it. Feel free to change the code of the testing framework accordingly.

About the floating point precision issues: We noted that behavior already. For float values, the testing framework already leverages [Almost.Equal()](https://github.com/prometheus/prometheus/blob/e92a18b6ce295b99e1c089d6e2bc62fbe3f73869/util/almost/almost.go#L22). It's much more complicated to do the same for native histograms, so we postponed that enhancement until we would run into floating point precision problems. That has happened now. Much appreciated if you would like to work on it.

**Comment by beorn7** [comment]
In different news: @NeerajGartia21 is also back in action, so maybe you could divide the work between the two of you or help each other with code reviews and such.

**Comment by fatsheep9146** [comment]
@NeerajGartia21 Welcome back! Right now I'm working on the `[TestNativeHistogram_Sum_Count_Add_AvgOperator](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L4193C6-L4193C51)` and `[TestNativeHistogram_SubOperator](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L4388C6-L4388C37)`, and both meet some problems about testing framework. And already submit the pr for TestNativeHistogram_Sum_Count_Add_AvgOperator, feel free to pick anyone you are interested in, I will help review the code.

**Comment by fatsheep9146** [comment]
> > So I think the actual results should also be compacted with Compact(0) before comparing.
> 
> Sounds like it. Feel free to change the code of the testing framework accordingly.
> 
> About the floating point precision issues: We noted that behavior already. For float values, the testing framework already leverages [Almost.Equal()](https://github.com/prometheus/prometheus/blob/e92a18b6ce295b99e1c089d6e2bc62fbe3f73869/util/almost/almost.go#L22). It's much more complicated to do the same for native histograms, so we postponed that enhancement until we would run into floating point precision problems. That has happened now. Much appreciated if you would like to work on it.

Alright! I will try to enhance the framework in the same pr. @beorn7

**Comment by NeerajGartia21** [comment]
> @NeerajGartia21 Welcome back! Right now I'm working on the [TestNativeHistogram_Sum_Count_Add_AvgOperator](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L4193C6-L4193C51) and [TestNativeHistogram_SubOperator](https://github.com/prometheus/prometheus/blob/8655fe5401c79e1fc176ca8b4644db04462dc9f5/promql/engine_test.go#L4388C6-L4388C37), and both meet some problems about testing framework. And already submit the pr for TestNativeHistogram_Sum_Count_Add_AvgOperator, feel free to pick anyone you are interested in, I will help review the code.

Cool! Thanks @fatsheep9146 ! I'll look into the last one then.

**Comment by NeerajGartia21** [comment]
@beorn7 @fatsheep9146 currently promql testing framework doesn't support Inf and NaN in histogram buckets. And in the tests for div we need this. Should I create a separate PR for this or do it in the same PR?

**Comment by beorn7** [comment]
Wasn't the plan that all buckets get removed if you divide by 0?

In any case, I guess Inf and even NaN _are_ values that can happen in a histogram, for example, I could just multiply a histogram with high bucket populations by a large number to cause `+Inf`. And then I could do the same to do `-Inf`. And then I could add the one histogram to the other to create `NaN` buckets.

So in short, yes, the testing framework should be able to create and test bucket populations that are ±Inf or NaN.

**Comment by Freathers** [comment]
> Wasn't the plan that all buckets get removed if you divide by 0?

Yes, that's the plan but since I'm replicating the current tests, and in the current one the bucket elements becomes Inf.

Also I've added Inf and NaN as numbers in histogram buckets in #14688 @beorn7 and @fatsheep9146 please review that.

**Comment by NeerajGartia21** [comment]
> Wasn't the plan that all buckets get removed if you divide by 0?

Yes, that's the plan but since I'm replicating the current tests, and in the current one the bucket elements becomes Inf.

Also I've added Inf and NaN as numbers in histogram buckets in https://github.com/prometheus/prometheus/pull/14688 @beorn7 and @fatsheep9146 please review that.

**Comment by beorn7** [comment]
I think we are done here. I'll close it, but if anyone finds a test that should be converted but isn't yet, please follow up.

Thanks, everyone involved, for all your work.
